torch_xla_py/xla_dist_test.py

```diff
@@ -128,15 +128,362 @@ def test_validate_diff_num_workers(self):
+      'metadata': {
+          'fingerprint': 'abc',
+          'items': [{
+              'key':
+                  'instance-template',
+              'value': ('projects/123456789012/global/'
+                        'instanceTemplates/fake-ig-template'),
+          }, {
+              'key':
+                  'created-by',
+              'value': ('projects/123456789012/zones/fake-zone/'
+                        'instanceGroupManagers/fake-ig'),
+          }],
+          'kind': 'compute#metadata',
+      },
+      'selfLink':
+          '{}/instances/{}'.format(project_zone_prefix, instance_name),
+      'networkInterfaces': [{
+          'networkIP': internal_ip,
+      }],
+      'status':
+          status,
+      'zone':
+          project_zone_prefix,
+  }
+
+
+def gen_fake_ig_list_instances_entry(instance_name, status):
+  project_zone_prefix = ('https://www.googleapis.com/compute/v1/'
+                         'projects/fake-project/zones/fake-zone')
+  return {
+      'instance': '{}/instances/{}'.format(project_zone_prefix, instance_name),
+      'status': status,
+  }
 
 
 class ClusterResolverTest(unittest.TestCase):
 
-  def test_cluster(self):
-    pass
+  def setUp(self):
+    super(ClusterResolverTest, self).setUp()
+    self.addCleanup(mock.patch.stopall)
+    mock.patch.object(ClusterResolver, '_get_instance_metadata',
+                      mock_request_metadata).start()
+    mock.patch.object(GoogleCredentials, 'get_application_default',
+                      lambda *args, **kwargs: None).start()
+    self.mock_discovery = mock.patch.object(
+        discovery, 'build', autospec=True).start()
+
+  def test_bad_empty_tpu_constructor(self):
+    tpus = []
+    self.assertRaisesRegex(ValueError, 'tpus must be a non-empty list',
+                           ClusterResolver, tpus)
+
+  def test_bad_none_tpu_constructor(self):
+    tpus = None
+    self.assertRaisesRegex(ValueError, 'tpus must be a non-empty list',
+                           ClusterResolver, tpus)
+
+  def test_bad_empty_vm_constructor(self):
+    tpus = ['fake-tpu']
+    vms = []
+    self.assertRaisesRegex(ValueError,
+                           'vms must be a non-empty list if provided',
+                           ClusterResolver, tpus, vms)
+
+  def test_healthy_instance_group_client_cluster(self):
+    # Arrange
+    list_instances_map = {
+        'fake-ig': {
+            'kind':
+                'compute#instanceGroupsListInstances',
+            'items': [
+                gen_fake_ig_list_instances_entry('fake-ig-a', 'RUNNING'),
+                gen_fake_ig_list_instances_entry('fake-ig-b', 'RUNNING'),
+                gen_fake_ig_list_instances_entry('fake-ig-c', 'RUNNING'),
+                gen_fake_ig_list_instances_entry('fake-ig-d', 'RUNNING'),
+            ],
+        },
+    }
+    instance_resp_map = {
+        'fake-ig-a':
+            gen_fake_instances_get_entry('fake-ig-a', 'n1-standard-16',
```

Comment:
similarly,

```python
{'fake-ig-'+char1: gen('fake-ig-a', 'n1', '10.0.0.'+char2, 'RUNNING')
 for char1, char2 in zip('abcd', '0123')}
```

Reply:
Done and also updated formatting for expected worker list to use similar formatting.